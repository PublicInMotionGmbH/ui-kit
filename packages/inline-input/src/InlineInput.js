import React from 'react'
import PropTypes from 'prop-types'

import { buildClassName } from '@talixo/shared'

export const moduleName = 'inline-input'

const propTypes = {
  /** Additional class name. */
  className: PropTypes.string,

  /** Prevents editing the content. */
  disabled: PropTypes.bool,

  /** Input value */
  value: PropTypes.string
}

const defaultProps = {
  disabled: false,
  value: ''
}

/**
 * Component which represents Inline Input.
 *
 * @property {object} props
 * @property {string} [props.className]
 * @property {boolean} [props.disabled]
 * @property {string} [props.value]
 * @class
 */
class InlineInput extends React.Component {
  state = {
    editing: false,
    inputValue: this.props.value
  }

  componentDidUpdate (prevProps, prevState) {
    if (prevState.editing !== this.state.editing && !prevState.editing && !this.props.disabled) {
      this.focusInput(this._input)
    }
  }

  focusInput = (obj) => {
    obj.focus()
  }

  handleInputChange = (e) => {
    const inputValue = e.target.value
    this.setState({ inputValue })
  }

  handleSpanClick = () => {
    this.setState({ editing: true })
  }

  handleBlur = () => {
    this.setState({ editing: false })
  }

  setRef = (node) => {
    this._input = node
  }

  render () {
    const { className, disabled, value, ...passedProps } = this.props
    const { editing, inputValue } = this.state

    const wrapperClsName = buildClassName(moduleName, className, { disabled })
    const inputClsName = buildClassName([moduleName, 'input'])
    const spanClsName = buildClassName([moduleName, 'span'], null, { disabled })

    return (
      <div className={wrapperClsName} {...passedProps}>
        {editing && !disabled
          ? <input
            className={inputClsName}
            ref={node => this.setRef(node)}
            placeholder={inputValue}
            onChange={(e) => this.handleInputChange(e)}
            onBlur={this.handleBlur}
            type='text'
          />
          : <span
            className={spanClsName}
            onClick={this.handleSpanClick}
          >
            {inputValue}
          </span>
        }
      </div>
    )
  }
}

InlineInput.propTypes = propTypes

InlineInput.defaultProps = defaultProps

export default InlineInput
