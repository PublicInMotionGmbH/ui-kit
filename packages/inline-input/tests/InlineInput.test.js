import React from 'react'
import { mount, shallow } from 'enzyme'

import InlineInput, { moduleName } from '../src/InlineInput'

const createWrapper = props => shallow(<InlineInput {...props} />)

describe('module name', () => {
  it('exports correctly', () => {
    expect(moduleName).toEqual('inline-input')
  })
})

describe('focusInput', () => {
  it('should focus', () => {
    const mock = { focus: jest.fn() }
    const wrapper = shallow(<InlineInput />)
    wrapper.instance().focusInput(mock)
    expect(mock.focus).toHaveBeenCalled()
  })
})

describe('<InlineInput />', () => {
  let span, wrapper
  beforeEach(() => {
    wrapper = createWrapper()
    span = wrapper.find('span')
  })

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('passes className to wrapper', () => {
    const className = 'new-class-name'
    const wrapper = createWrapper({ className: className })
    expect(wrapper.props().className.includes(className)).toEqual(true)
  })

  it('renders disabled correctly', () => {
    const wrapper = createWrapper({ disabled: true })
    expect(wrapper.props().className.includes('disabled')).toEqual(true)
  })

  it('should not change to input when disabled is set to true', () => {
    const wrapper = createWrapper({ disabled: true })
    span.simulate('click')
    expect(wrapper.find('span').exists()).toEqual(true)
    expect(wrapper.find('input').exists()).toEqual(false)
  })

  it('should change state.editing to true when span is clicked', () => {
    span.simulate('click')
    expect(wrapper.state().editing).toEqual(true)
  })

  it('should change span to input when span is clicked', () => {
    span.simulate('click')
    expect(wrapper.find('span').exists()).toEqual(false)
    expect(wrapper.find('input').exists()).toEqual(true)
  })

  it('should change state.inputValue when typing inside input', () => {
    span.simulate('click')
    const input = wrapper.find('input')
    input.simulate('change', { target: { value: 'name' } })
    expect(wrapper.state().inputValue).toEqual('name')
  })

  it('should change to span on blur', () => {
    span.simulate('click')
    const input = wrapper.find('input')
    input.simulate('blur')
    expect(wrapper.find('span').exists()).toEqual(true)
    expect(wrapper.find('input').exists()).toEqual(false)
  })

  it('gets input ref correctly', () => {
    const wrapper = mount(<InlineInput />)
    const span = wrapper.find('span')
    span.simulate('click')
    const input = wrapper.find('input')
    wrapper.instance().setRef(input)
    expect(wrapper.instance()._input).toEqual(input)
    wrapper.unmount()
  })
})
