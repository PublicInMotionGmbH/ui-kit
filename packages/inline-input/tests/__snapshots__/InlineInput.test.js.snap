// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InlineInput /> renders correctly 1`] = `
<div
  className="talixo-inline-input"
>
  <span
    className="talixo-inline-input__span"
    onClick={[Function]}
  />
</div>
`;
